SimTracker/TrackAssociation/plugins/TrackingParticleSelectorByGen.cc

```diff
@@ -0,0 +1,341 @@
+      allIndices.push_back(index);
+      if (cachedIndex != index) {
+        flags_[index] = keepOrDrop;
+        recursiveFlagDaughters(index, src, keepOrDrop, allIndices);
+      }
+    }
+  }
+}
+
+void TrackingParticleSelectorByGen::recursiveFlagMothers(size_t index,
+                                                         const reco::GenParticleCollection &src,
+                                                         int keepOrDrop,
+                                                         std::vector<size_t> &allIndices) {
+  GenParticleRefVector mothers = src[index].motherRefVector();
+  // avoid infinite recursion if the mothers are set to "this" particle.
+  size_t cachedIndex = index;
+  for (GenParticleRefVector::const_iterator i = mothers.begin(); i != mothers.end(); ++i) {
+    index = i->key();
+    // To also avoid infinite recursion if a "loop" is found in the daughter list,
+    // check to make sure the index hasn't already been added.
+    if (find(allIndices.begin(), allIndices.end(), index) == allIndices.end()) {
+      allIndices.push_back(index);
+      if (cachedIndex != index) {
+        flags_[index] = keepOrDrop;
+        recursiveFlagMothers(index, src, keepOrDrop, allIndices);
+      }
+    }
+  }
+}
+
+void TrackingParticleSelectorByGen::getDaughterKeys(vector<size_t> &daIndxs,
+                                                    vector<size_t> &daNewIndxs,
+                                                    const GenParticleRefVector &daughters) const {
+  for (GenParticleRefVector::const_iterator j = daughters.begin(); j != daughters.end(); ++j) {
+    GenParticleRef dau = *j;
+    if (find(daIndxs.begin(), daIndxs.end(), dau.key()) == daIndxs.end()) {
+      daIndxs.push_back(dau.key());
+      int idx = flags_[dau.key()];
+      if (idx > 0) {
+        daNewIndxs.push_back(idx);
+      } else {
+        const GenParticleRefVector &daus = dau->daughterRefVector();
+        if (!daus.empty())
+          getDaughterKeys(daIndxs, daNewIndxs, daus);
+      }
+    }
+  }
+}
+
+void TrackingParticleSelectorByGen::getMotherKeys(vector<size_t> &moIndxs,
+                                                  vector<size_t> &moNewIndxs,
+                                                  const GenParticleRefVector &mothers) const {
+  for (GenParticleRefVector::const_iterator j = mothers.begin(); j != mothers.end(); ++j) {
+    GenParticleRef mom = *j;
+    if (find(moIndxs.begin(), moIndxs.end(), mom.key()) == moIndxs.end()) {
+      moIndxs.push_back(mom.key());
+      int idx = flags_[mom.key()];
+      if (idx >= 0) {
+        moNewIndxs.push_back(idx);
+      } else {
+        const GenParticleRefVector &moms = mom->motherRefVector();
+        if (!moms.empty())
+          getMotherKeys(moIndxs, moNewIndxs, moms);
+      }
+    }
+  }
+}
+
+void TrackingParticleSelectorByGen::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  using namespace edm;
+
+  if (firstEvent_) {
+    PdgEntryReplacer rep(iSetup);
```

Comment:
This needs to be migrated to EventSetup consumes. (I understand it would have implications outside of the code added in this PR, so this maybe a note more for @cms-sw/simulation-l2)